### PR TITLE
[objects pruner] add batching

### PR DESCRIPTION
**Pull request by phoenix-o**
initial version didn't manage to keep up with narwhal
adding batching to resolve the issue

### CODE REVIEW

File: crates/sui-core/src/authority/authority_store_pruner.rs [hunk 1]
```diff
@@ -18,13 +19,16 @@ use tokio::{
...
 };
+use tracing::debug;
 use tracing::log::{error, info};
 use typed_store::rocks::DBMap;
 use typed_store::Map;
 
 use super::authority_store_tables::AuthorityPerpetualTables;
 
 const MAX_OPS_IN_ONE_WRITE_BATCH: u64 = 10000;
+const MAX_TRANSACTIONS_IN_ONE_BATCH: usize = 1000;
```
Comment (phoenix-o) [inline crates/sui-core/src/authority/authority_store_pruner.rs:30]: should we add those parameters into configuration instead?

File: crates/sui-core/src/authority/authority_store_pruner.rs [hunk 2]
```diff
@@ -8,126 +8,26 @@ use std::collections::HashMap;
...
-            }
-        }
-        info!(
-            "Finished compacting objects, keys scanned = {:?}, objects scanned = {:?}",
-            total_keys_scanned, total_objects_scanned
-        );
-        total_pruned
-    }
-
     fn handle_checkpoint(
```
Comment (gdanezis) [inline crates/sui-core/src/authority/authority_store_pruner.rs:31]: I know this is not a place where this PR modifies things but it should:
* Again just a small doc to tell us what this fn should do, `handle_checkpoint` really means too little.
* The `skip_prior_to` operation is expensive. It is safe to use as the first key (object_id, 0) and last key (object_id, version.inc()).
* The use of `delete_range` is a problem, since it call `delete_range_cf`, and the docs specifically say you should not do so for many small ranges (see: https://docs.rs/rocks/latest/rocks/db/struct.DB.html#method.delete_range_cf).
* Using an iterator over the DB as in `iter.take_while` purely to compute `pruned` which I suspect is just a metric is a pretty expensive thing to do. (Which is then ignored in many places.)

File: crates/sui-core/src/authority/authority_store_pruner.rs [hunk 3]
```diff
@@ -160,25 +60,41 @@ impl AuthorityStorePruner {
...
             .unwrap_or_default();
+        let mut checkpoints_in_batch = 0;
+        let mut batch_effects = vec![];
+        let mut network_total_transactions = 0;
+
+        debug!(
+            "Starting object pruning. Current epoch: {}. Latest pruned checkpoint: {}",
+            current_epoch, checkpoint_number
+        );
         loop {
```
Comment (gdanezis) [inline crates/sui-core/src/authority/authority_store_pruner.rs:84]: I think this loop tries to get all checkpoints for past epoch. I am sure (at least we should) at this point have a table of epoch -> checkpoint number we can use to identify this range, and then use an iterator and multi-get to get all checkpoints at once? Doing one by one in a blocking fashion is not very databasey for such a bulk operation.

File: crates/sui-core/src/authority/authority_store_pruner.rs [hunk 4]
```diff
@@ -189,53 +105,51 @@ impl AuthorityStorePruner {
...
-                            info!("Completed flushing objects table");
-                        } else {
-                            error!("Failed to flush objects table");
-                        }
-                    },
-                    _ = live_prune_interval.tick(), if num_epochs_to_retain != u64::MAX => {
-                        match Self::process_checkpoints(&perpetual_db, &checkpoint_store, num_epochs_to_retain) {
-                            Ok(()) => info!("Pruned checkpoints"),
-                            Err(err) => error!("Failed to prune objects: {:?}", err),
+                    _ = prune_interval.tick(), if num_epochs_to_retain != u64::MAX => {
```
Comment (gdanezis) [inline crates/sui-core/src/authority/authority_store_pruner.rs:150]: I think this interval should be in hours not milliseconds, since my understanding of pruning is that it deletes past epochs, which are unlikely to have changed every few millis.

File: crates/sui-core/src/authority/authority_store_pruner.rs [hunk 5]
```diff
@@ -189,53 +105,51 @@ impl AuthorityStorePruner {
...
+                Self::handle_checkpoint(batch_effects, &perpetual_db.objects)?;
+                checkpoint_store
+                    .update_highest_pruned_checkpoint(checkpoint_number, checkpoint_digest)?;
+                batch_effects = vec![];
+                checkpoints_in_batch = 0;
+            }
+        }
+        if !batch_effects.is_empty() {
+            Self::handle_checkpoint(batch_effects, &perpetual_db.objects)?;
+            checkpoint_store
```
Comment (gdanezis) [inline crates/sui-core/src/authority/authority_store_pruner.rs:122]: Can we afford to update this at the end? What is the worse that happens? We crash before updating, and try to prune again? Is this a problem?

File: crates/sui-core/src/authority/authority_store_pruner.rs [hunk 6]
```diff
@@ -189,53 +105,51 @@ impl AuthorityStorePruner {
...
-                        } else {
-                            error!("Failed to flush objects table");
-                        }
-                    },
-                    _ = live_prune_interval.tick(), if num_epochs_to_retain != u64::MAX => {
-                        match Self::process_checkpoints(&perpetual_db, &checkpoint_store, num_epochs_to_retain) {
-                            Ok(()) => info!("Pruned checkpoints"),
-                            Err(err) => error!("Failed to prune objects: {:?}", err),
+                    _ = prune_interval.tick(), if num_epochs_to_retain != u64::MAX => {
+                        if let Err(err) = Self::prune_objects(&perpetual_db, &checkpoint_store, num_epochs_to_retain) {
```
Comment (gdanezis) [inline crates/sui-core/src/authority/authority_store_pruner.rs:151]: This is a heavily blocking task, that may take a very long time. We should probably spin it on a `spawn_blocking` task.

**Review by gdanezis — COMMENTED** [review]
Given that this PR is about increasing efficiency I had a look and left quite a few pointers.

File: crates/sui-core/src/authority/authority_store_pruner.rs [hunk 7]
```diff
@@ -160,25 +60,41 @@ impl AuthorityStorePruner {
...
+        debug!(
+            "Starting object pruning. Current epoch: {}. Latest pruned checkpoint: {}",
+            current_epoch, checkpoint_number
+        );
         loop {
-            let Some(checkpoint) = checkpoint_store.get_checkpoint_by_sequence_number(pruned_seq_number + 1)?
-                else {return Ok(());};
+            let Some(checkpoint) = checkpoint_store.get_checkpoint_by_sequence_number(checkpoint_number + 1)? else {break;};
             // checkpoint's epoch is too new. Skipping for now
             if current_epoch < checkpoint.epoch() + num_epochs_to_retain {
```
[the remark this answers is not in the document]
Comment (phoenix-o) [reply crates/sui-core/src/authority/authority_store_pruner.rs:87]: kind of. Code currently supports num_epochs_to_retain = 0, which implies aggressive pruner that tries to delete objects as soon as it can.
Code can be somewhat simplified if we agree not to support such behaviour and always keep at least one extra epoch of data

File: crates/sui-core/src/authority/authority_store_pruner.rs [hunk 8]
```diff
@@ -160,25 +60,41 @@ impl AuthorityStorePruner {
...
-                return Ok(());
+                break;
+            }
+            checkpoint_number = checkpoint.sequence_number();
+            checkpoint_digest = checkpoint.digest();
+            checkpoints_in_batch += 1;
+            if network_total_transactions == checkpoint.summary.network_total_transactions {
```
[the remark this answers is not in the document]
Comment (phoenix-o) [reply crates/sui-core/src/authority/authority_store_pruner.rs:93]: it tries to skip processing empty checkpoints to avoid any extra rocksdb IO operations
please let me know if there's a better way to verify that checkpoint is not empty

File: crates/sui-core/src/authority/authority_store_pruner.rs (hunk 2 above)
Comment (phoenix-o) [reply crates/sui-core/src/authority/authority_store_pruner.rs:31]: > handle_checkpoint really means too little

agree, will somewhat change naming
>  delete_range concern
rocksdb stores delete_range intervals in the separate DS. Multiple delete_ranges might indeed be an issue. We observed it in a past. For example, during `get` operation rocksdb will typically try to apply and merge them to return the result. That's why we're currently passing `ignore_range_deletions` flag to objects. So now deleted ranges are stored separately and completely ignored by reads. Basically they are only visible for compaction job to permanently remove the data
We ran experiment on private-testnet with current PR and we don't see any issues.
For context, as a followup, we're still trying to migrate from delete_range to point-delete API. It will somewhat simplify this function and will allow to enable transactions on given rocksdb instance. The concern is that point-deletes in rocksdb are creating tombstones, thus iterator seek performance might be impacted. We're running experiments now to validate that
> take_while concern
it's used not for a metric, but to form a corresponding delete range. But it really might be unnecessary

File: crates/sui-core/src/authority/authority_store_pruner.rs (hunk 3 above)
Comment (phoenix-o) [reply crates/sui-core/src/authority/authority_store_pruner.rs:84]: good point. It's done mostly to support case of num_epochs_to_retain = 0. We're doing batching on application layer, which is suboptimal. We can apply this optimization later if we get rid of option of aggressive pruner

File: crates/sui-core/src/authority/authority_store_pruner.rs (hunk 4 above)
Comment (phoenix-o) [reply crates/sui-core/src/authority/authority_store_pruner.rs:150]: same, related to `num_epochs_to_retain = 0`

File: crates/sui-core/src/authority/authority_store_pruner.rs (hunk 5 above)
Comment (phoenix-o) [reply crates/sui-core/src/authority/authority_store_pruner.rs:122]: pruning should be idempotent. We update as we go once we fill the batch or at the end of the loop if there are no more checkpoints to process in the current epoch

File: crates/sui-core/src/authority/authority_store_pruner.rs [hunk 9]
```diff
@@ -189,53 +104,51 @@ impl AuthorityStorePruner {
             if effects.iter().any(|effect| effect.is_none()) {
                 return Err(anyhow::anyhow!("transaction effects data is missing"));
             }
-            Self::handle_checkpoint(effects.into_iter().flatten(), &perpetual_db.objects)?;
-            checkpoint_store.update_highest_pruned_checkpoint(&checkpoint)?;
-            pruned_seq_number += 1;
+            batch_effects.extend(effects.into_iter().flatten());
+
```
Comment (sadhansood) [inline crates/sui-core/src/authority/authority_store_pruner.rs:108]: If we have `prune_effects` return `HashMap<ObjectID, (VersionNumber, VersionNumber)>` we won't need to batch effects in memory and can do a bulk rocksdb delete. Right now we are bound to invoke delete across fewer effects since we probably can't keep many of those in memory

File: crates/sui-core/src/authority/authority_store_pruner.rs [hunk 10]
```diff
@@ -401,205 +268,4 @@ mod tests {
...
-            before_compaction_size, after_compaction_size
-        );
-        ma::assert_le!(after_compaction_size, before_compaction_size);
-        Ok(())
-    }
-
-    #[cfg(not(target_env = "msvc"))]
-    #[tokio::test]
-    async fn ensure_tombstone_fragmentation_in_stack_frame_without_ignore_range_delete(
-    ) -> Result<(), anyhow::Error> {
```
Comment (sadhansood) [inline crates/sui-core/src/authority/authority_store_pruner.rs:514]: Some of these tests are relevant as long as we have `delete_range` and ignore_delete_range in the code path

**Review by sadhansood — APPROVED** [review]
